unit_testing/LightController/LightController_tests.cpp

```diff
@@ -0,0 +1,432 @@
+    };
+
+    auto expectStmNotNull = [](StmGpio* stm) {
+        EXPECT_NE(stm->set,    nullptr);
+        EXPECT_NE(stm->get,    nullptr);
+        EXPECT_NE(stm->toggle, nullptr);
+    };
+
+    /* The GPIOs should be null prior to init function */
+    for(int i = 0; i < LED_CHANNELS*NO_COLORS; i++) expectStmNull(ChCtrl[i]);
+
+    LightControllerInit(&htim2, &htim5, &hwwdg);
+
+    /* The GPIOs should be instatiated after init function */
+    for(int i = 0; i < LED_CHANNELS*NO_COLORS; i++) expectStmNotNull(ChCtrl[i]);
+
+    /* The GPIOs should be off after init */
+    for(int i = 0; i < LED_CHANNELS*NO_COLORS; i++) EXPECT_FALSE(stmGetGpio(*ChCtrl[i]));
+}
+
+TEST_F(LightControllerTest, InputHandler) 
+{
+    LightControllerInit(&htim2, &htim5, &hwwdg);
+    /* Note: usb RX buffer is flushed during the first loop, so a single loop must be done before
+    ** printing anything */
+    LightControllerLoop(bootMsg);
+    HAL_TIM_PeriodElapsedCallback(&htim5);
+
+    // Initial flush
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "Boot Unit Test\r",
+        "0, 0, 0, 0x0"
+    ));
+
+
+    /* ----- VALID INPUTS ----- */
+    writeLightControllerMessage("p1 ABAB10\n");
+    HAL_TIM_PeriodElapsedCallback(&htim5);
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "abab10, 0, 0, 0x1"
+    ));
+
+    writeLightControllerMessage("p2 105020\n");
+    HAL_TIM_PeriodElapsedCallback(&htim5);
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "abab10, 105020, 0, 0x3"
+    ));
+
+    writeLightControllerMessage("p3 FFFFFF\n");
+    HAL_TIM_PeriodElapsedCallback(&htim5);
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "abab10, 105020, ffffff, 0x7"
+    ));
+
+    // Reset all ports
+    writeLightControllerMessage("p1 000000\n");
+    writeLightControllerMessage("p2 000000\n");
+    writeLightControllerMessage("p3 000000\n");
+    HAL_TIM_PeriodElapsedCallback(&htim5);
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "0, 0, 0, 0x0"
+    ));
+
+    /* ----- INVALID INPUTS ----- */
+
+    /* For the misreads we do not need to call the printtimer callback
+       since the MISREAD messages are pushed to the USB by the standard
+       protocol.*/
+
+    // Invalid port   
+    writeLightControllerMessage("p4 FFFFFF\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p4 FFFFFF"
+    ));
+
+    writeLightControllerMessage("p0 FFFFFF\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p0 FFFFFF"
+    ));
+
+    // Wrong hex code (too short)
+    writeLightControllerMessage("p1 13908\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p1 13908"
+    ));
+
+    // Wrong hex code (too long)
+    writeLightControllerMessage("p1 13908CC\n");
+    EXPECT_FLUSH_USB(ElementsAre(
+        "\r",
+        "MISREAD: p1 13908CC"
+    ));
+
+    // Wrong hex code (non valid hex characters)
+    writeLightControllerMessage("p1 ACACFM\n");
```

Comment:
I guess a common thing might be to write **0x**123456, should that be allowed?

Reply:
I'm having a hard time actually getting this working. Since loopcode never uses the 0x prefix I will leave it a later exercise.